File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, origNums: List[int]) -> List[List[int]]:
        nums = []
        valDict = {}
        for num in origNums:
            if num in valDict:
                valDict[num] += 1
            else:
                valDict[num] = 1

            if 3 > valDict[num] or 4 > valDict[num] and 0 == num:
                nums.append(num) 

        print(nums, valDict)

        triples = set()
        for i in range(len(nums)):
            valDict[nums[i]] -= 1
            for j in range(len(nums)):
                if i == j or nums[i] > nums[j]:
                    continue

                valDict[nums[j]] -= 1
                val = -nums[i] - nums[j]
                if nums[j] <= val and val in valDict and 0 < valDict[val]:
                    triples.add((nums[i], nums[j], val))
                valDict[nums[j]] += 1
            valDict[nums[i]] += 1

        return list(triples)

        '''
        dict = {}
        #nums.sort()
        for i in range(len(nums)):
            dict[nums[i]] = i

        triples = set()
        for i in range(len(nums)):
            for j in range(len(nums)):
                if j == i or nums[j] < nums[i]:
                    continue
                
                k = dict.get(-nums[i] - nums[j])
                if None == k:
                    continue

                if k != i and k != j and nums[k] >= nums[j]:
                    triples.add((nums[i], nums[j], nums[k]))

        return list(triples)
        '''

        '''
        triples = set()
        nums.sort()
        for i in range(len(nums) - 1):
            j = i + 1
            k = len(nums) - 1
            while j < k:
                if 0 > nums[i] + nums[j] + nums[k]:
                    j += 1
                elif 0 < nums[i] + nums[j] + nums[k]:
                    k -= 1
                else:
                    triples.add((nums[i], nums[j], nums[k]))
                    j += 1
                    k -= 1
    
        return list(triples)
        '''

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> triples = new ArrayList();
        Arrays.sort(nums);
        /*
        boolean[] choosen = new boolean[nums.length];
        for (int i = 0; i < nums.length - 2; i++) {
            if (choosen[i]) {
                break;
            }

            if (0 < i && nums[i - 1] == nums[i]) {
                continue;
            }

            HashMap<Integer, int[]> map = new HashMap();
            int j;
            for (j = i + 1; j < nums.length && !map.containsKey(nums[j]); j++) {
                if (choosen[j]) {
                    break;
                }

                if (i + 1 < j && nums[j - 1] == nums[j]) {
                    continue;
                }

                map.put(-(nums[i] + nums[j]), new int[]{nums[i], nums[j]});
            }

            int from = j;
            for (; j < nums.length; j++) {
                if (from < j && nums[j - 1] == nums[j]) {
                    continue;
                }

                if (map.containsKey(nums[j])) {
                    choosen[j] = true;
                    int[] pairs = map.get(nums[j]);
                    List<Integer> triple = new ArrayList();
                    triple.add(pairs[0]);
                    triple.add(pairs[1]);
                    triple.add(nums[j]);
                    triples.add(triple);
                }
            }
        }
        */

        for (int i = 0; i < nums.length - 2; i++) {
            if (0 < i && nums[i - 1] == nums[i]) {
                continue;
            }

            HashMap<Integer, Integer> map = new HashMap();
            int j;
            for (j = i + 1; j < nums.length - 1 && !map.containsKey(nums[j]); j++) {
                if (i + 1 < j && nums[j - 1] == nums[j]) {
                    continue;
                }

                map.put(-nums[i] - nums[j], nums[j]);
            }

            for (int k = j; k < nums.length; k++) {
                if (j < k && nums[k - 1] == nums[k]) {
                    continue;
                }

                if (map.containsKey(nums[k])) {
                    List<Integer> triple = new ArrayList();
                    triple.add(nums[i]);
                    triple.add(map.get(nums[k]));
                    triple.add(nums[k]);
                    triples.add(triple);
                }
            }
        }

        return triples;
    }
}